packages/realm/src/PropertyHelpers.ts

```diff
@@ -121,6 +125,45 @@ function embeddedSet({ typeHelpers: { toBinding }, columnKey }: PropertyOptions)
 type AccessorFactory = (options: PropertyOptions) => PropertyAccessor;
 
 const ACCESSOR_FACTORIES: Partial<Record<binding.PropertyType, AccessorFactory>> = {
+  [binding.PropertyType.Int](options) {
+    const { realm, columnKey, presentation, optional } = options;
+
+    if (presentation === "counter") {
+      return {
+        get(obj) {
+          return obj.getAny(columnKey) === null ? null : new Counter(realm, obj, columnKey);
+        },
+        set(obj, value, isCreating) {
+          // We only allow resetting a counter this way (e.g. realmObject.counter = 5)
+          // when it is first created, or when a nullable/optional counter was
+          // previously `null`, or when resetting a nullable counter to `null`.
+          if (isCreating) {
+            defaultSet(options)(obj, value);
+            return;
+          }
+
+          if (optional) {
+            const isUninitialized = obj.getAny(columnKey) === null;
+            const resettingToNull = value === null || value === undefined;
+            if (isUninitialized || resettingToNull) {
+              defaultSet(options)(obj, value);
+              return;
+            }
+          }
+          throw new Error(
+            "You can only reset a Counter instance when initializing a previously " +
+              "null Counter or resetting a nullable Counter to null. To update the " +
+              "value of the Counter, use its instance methods.",
+          );
```

Comment:
Not blocking, but I'm not a super-fan of duplicating `defaultSet(options)(obj, value)` here - is there a way to isolate the logic to determine if a set is allowed and have calling `defaultSet` depend on just that?

Reply:
Do you mean calling `defaultSet()` only in one place here? One thing I want to avoid is unnecessarily calling into Core (`obj.getAny()`).

Reply:
> calling `defaultSet()` only in one place here?

Right.

> avoid is unnecessarily calling into Core

That makes sense, but as long as the call into core is on the right-hand-side of a logical operator, it knows not to evaluate it, if it's not needed. Notice how this doesn't throw:

```js
> function foo() { throw new Error("!"); }
undefined
> const isAllowed = 2 > 1 || foo();
undefined
> isAllowed
true
```

Personally I would probably either make a local `const isAllowed` variable or define a separate `function isAllowed(obj, isCreating)` if that expression becomes too complex.

Reply:
Gotcha, yeah with the Core part I was referring more to keeping some of the local variables for readability that sometimes require evaluating it beforehand. The logic could still be expressed in one variable:

```ts
set(obj, value, isCreating) {
  // We only allow resetting a counter this way (e.g. realmObject.counter = 5)
  // when it is first created, or when a nullable/optional counter was
  // previously `null`, or when resetting a nullable counter to `null`.
  const isAllowed =
    isCreating || (optional && (value === null || value === undefined || obj.getAny(columnKey) === null));

  if (isAllowed) {
    defaultSet(options)(obj, value);
  } else {
    throw new Error(
      "You can only reset a Counter instance when initializing a previously " +
        "null Counter or resetting a nullable Counter to null. To update the " +
        "value of the Counter, use its instance methods.",
    );
  }
},
```